### PR TITLE
the first test-version of virtual cluster

### PR DESCRIPTION
**Pull request by charleszheng44**
how to create a virtual cluster:

```bash
cd incubator/virtualcluster
kubectl apply -f config/crds
```
in another terminal run 

```bash
go run ./cmd/manager/main.go
```
back to the first terminal, and run
```bash
kubectl apply -f config/sampleswithspec/clusterversion_v1.yaml
kubectl apply -f config/sampleswithspec/virtualcluster.yaml
```
you will see three pod running in the `vc-sample` namespace in few minutes

```bash
NAME                   READY   STATUS    RESTARTS   AGE
apiserver-0            1/1     Running   0          24m
controller-manager-0   1/1     Running   0          24m
etcd-0                 1/1     Running   0          24m
```

### CODE REVIEW

File: incubator/virtualcluster/config/sampleswithspec/clusterversion_v1.yaml [hunk 1]
```diff
@@ -9,87 +9,94 @@ spec:
...
-                    - -ec
-                    - ETCDCTL_API=3 etcdctl --endpoints=etcd.vc-test.svc.cluster.local --cacert=/etc/kubernetes/pki/etcd/ca.pem --cert=/etc/kubernetes/pki/etcd/healthcheck-client.crt --key=/etc/kubernetes/pki/etcd/healthcheck-client.key get foo 
-                  failureThreshold: 8
-                  initialDelaySeconds: 60
-                  timeoutSeconds: 15
-                volumeMounts:
-                - mountPath: /etc/kubernetes/pki/etcd
-                  name: etcd-ca
+            - name: etcd
+              image: chaozheng/etcd-v3.4.0-amd64
```
Comment (zhuangqh) [inline incubator/virtualcluster/config/sampleswithspec/clusterversion_v1.yaml]: using a public registry?  registry.hub.docker.com/virtualcluster

File: incubator/virtualcluster/pkg/controller/pki/pki.go [hunk 2]
```diff
@@ -110,17 +110,19 @@ func NewAPIServerKubeletClientCertAndKey(ca *CrtKeyPair) (*x509.Certificate, *rs
 }
 
 // NewEtcdServerCrtAndKey creates new crt-key pair using ca for etcd
-func NewEtcdServerCrtAndKey(ca *CrtKeyPair, etcdDomain string) (*CrtKeyPair, error) {
-
+func NewEtcdServerCrtAndKey(ca *CrtKeyPair, etcdDomains []string) (*CrtKeyPair, error) {
 	// create AltNames with defaults DNSNames/IPs
+	// log.Println("adding etcd domains into cert ", "domains ", etcdDomains)
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/controller/pki/pki.go]: Can we remove this uncomment line? or change it to a debug log.

File: incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go [hunk 3]
```diff
@@ -185,28 +185,88 @@ func (r *ReconcileVirtualcluster) createPKI(vc *tenancyv1alpha1.Virtualcluster,
 	return nil
 }
 
+// genInitialClusterArgs generates the values for `--inital-cluster` option of etcd based on the number of
+// replicas specified in etcd StatefulSet
+func genInitialClusterArgs(replicas int32, stsName, svcName string) (argsVal string) {
+	for i := int32(0); i < replicas; i++ {
+		peerAddr := fmt.Sprintf("%s-%d=https://%s-%d.%s:2380", stsName, i, stsName, i, svcName)
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go]: I am confused about the `2380` here

File: incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go [hunk 4]
```diff
@@ -185,28 +185,88 @@ func (r *ReconcileVirtualcluster) createPKI(vc *tenancyv1alpha1.Virtualcluster,
...
+		}
+		argsVal = argsVal + peerAddr + ","
+	}
+
+	return argsVal
+}
+
+// completmentETCDTemplate completments the ETCD template of the specified clusterversion
+// based on the virtual cluster setting
+func completmentETCDTemplate(vcName string, etcdBdl *tenancyv1alpha1.StatefulSetSvcBundle) {
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go]: s/completment/complement/

File: incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go [hunk 5]
```diff
@@ -185,28 +185,88 @@ func (r *ReconcileVirtualcluster) createPKI(vc *tenancyv1alpha1.Virtualcluster,
...
+	etcdBdl.Service.ObjectMeta.Namespace = vcName
+	args := &etcdBdl.StatefulSet.Spec.Template.Spec.Containers[0].Args
+	icaVal := genInitialClusterArgs(*etcdBdl.StatefulSet.Spec.Replicas,
+		etcdBdl.StatefulSet.Name, etcdBdl.Service.Name)
+	*args = append(*args, "--initial-cluster", icaVal)
+}
+
+// completmentETCDTemplate completments the apiserver template of the specified clusterversion
+// based on the virtual cluster setting
+func completmentAPIServerTemplat(vcName string, apiserverBdl *tenancyv1alpha1.StatefulSetSvcBundle) {
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go]: s/completmentAPIServerTemplat/complementAPIServerTemplate/

File: incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go [hunk 6]
```diff
@@ -185,28 +185,88 @@ func (r *ReconcileVirtualcluster) createPKI(vc *tenancyv1alpha1.Virtualcluster,
...
+func completmentETCDTemplate(vcName string, etcdBdl *tenancyv1alpha1.StatefulSetSvcBundle) {
+	etcdBdl.StatefulSet.ObjectMeta.Namespace = vcName
+	etcdBdl.Service.ObjectMeta.Namespace = vcName
+	args := &etcdBdl.StatefulSet.Spec.Template.Spec.Containers[0].Args
+	icaVal := genInitialClusterArgs(*etcdBdl.StatefulSet.Spec.Replicas,
+		etcdBdl.StatefulSet.Name, etcdBdl.Service.Name)
+	*args = append(*args, "--initial-cluster", icaVal)
+}
+
+// completmentETCDTemplate completments the apiserver template of the specified clusterversion
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go]: s/completmentETCDTemplate/complementAPIServerTemplate/
please keep the comment sync with code😄

File: incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go [hunk 7]
```diff
@@ -185,28 +185,88 @@ func (r *ReconcileVirtualcluster) createPKI(vc *tenancyv1alpha1.Virtualcluster,
...
+
+// completmentETCDTemplate completments the ETCD template of the specified clusterversion
+// based on the virtual cluster setting
+func completmentETCDTemplate(vcName string, etcdBdl *tenancyv1alpha1.StatefulSetSvcBundle) {
+	etcdBdl.StatefulSet.ObjectMeta.Namespace = vcName
+	etcdBdl.Service.ObjectMeta.Namespace = vcName
+	args := &etcdBdl.StatefulSet.Spec.Template.Spec.Containers[0].Args
+	icaVal := genInitialClusterArgs(*etcdBdl.StatefulSet.Spec.Replicas,
+		etcdBdl.StatefulSet.Name, etcdBdl.Service.Name)
+	*args = append(*args, "--initial-cluster", icaVal)
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go]: this code snippet is quite c++ style..
how about in go style
```go
func completmentETCDTemplate(vcName string, etcdBdl *tenancyv1alpha1.StatefulSetSvcBundle) {
	etcdBdl.StatefulSet.ObjectMeta.Namespace = vcName
	etcdBdl.Service.ObjectMeta.Namespace = vcName
	args := etcdBdl.StatefulSet.Spec.Template.Spec.Containers[0].Args
	icaVal := genInitialClusterArgs(*etcdBdl.StatefulSet.Spec.Replicas,
		etcdBdl.StatefulSet.Name, etcdBdl.Service.Name)
	args = append(args, "--initial-cluster", icaVal)
	etcdBdl.StatefulSet.Spec.Template.Spec.Containers[0].Args = args
}

```

File: incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go [hunk 8]
```diff
@@ -235,30 +295,49 @@ func (r *ReconcileVirtualcluster) createPKISecrets(caGroup *vcpki.ClusterCAGroup
...
+}
+
+// createVirtualcluster creates a new Virtualcluster based on the specified ClusterVersion
 func (r *ReconcileVirtualcluster) createVirtualcluster(vc *tenancyv1alpha1.Virtualcluster, cv *tenancyv1alpha1.ClusterVersion) (reconcile.Result, error) {
 	// 1. create ns
-	if err := r.createNamespace(vc.Name); err != nil {
-		return reconcile.Result{}, err
+	err := r.createNamespace(vc.Name)
+	if err != nil {
+		goto Cleanup
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go]: we could use `defer` to do the cleanup job.
take https://github.com/alibaba/pouch/blob/master/ctrd/container.go#L692 for reference.

File: incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go (hunk 3 above)
Comment (charleszheng44) [reply incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go]: 2380 is the default port for etcd peer to communicate with each other.

File: incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go (hunk 5 above)
Comment (charleszheng44) [reply incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go]: good catch

File: incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go (hunk 3 above)
Comment (charleszheng44) [reply incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go]: assign the default port number to a constant

File: incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go (hunk 7 above)
Comment (charleszheng44) [reply incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go]: change to go style

File: incubator/virtualcluster/pkg/controller/pki/pki.go (hunk 2 above)
Comment (charleszheng44) [reply incubator/virtualcluster/pkg/controller/pki/pki.go]: useless comment removed

**Pull request closed without merging**